File: Week_1/OOPS/library.cpp
#include<bits/stdc++.h>
using namespace std;
class LibraryUser{
    public:
        virtual void registerAccount()=0;
        virtual void requestBook()=0;
    virtual ~LibraryUser() = default;
};


class KidUsers: public LibraryUser{
    public:
        int age;
        string bookType;
    public:
        void registerAccount() override{
            if(age<=12){
                cout << "You have successfully registered under a Kids Account" << endl;
            }
            else if(age>12){
                cout << "Sorry, Age must be less than 12 to register as a kid" << endl;
            }
        }

        void requestBook() override{
            if(bookType=="Kids"){
                cout << "Book Issued successfully, please return the book within 10 days" << endl;
            }
            else{
                cout << "Oops, you are allowed to take only kids books" << endl;
            }
        }
};

class AdultUsers: public LibraryUser{
    public:
        int age;
        string bookType;
    public:
        void registerAccount() override{
            if(age>12){
                cout << "You have successfully registered under an Adult Account" << endl;
            }
            else if(age<=12){
                cout << "Sorry, Age must be greater than 12 to register as an adult" << endl;
            }
        }

        void requestBook() override{
            if(bookType=="Fiction"){
                cout << "Book issued successfully, please return the book within 7 days" << endl;
            }
            else{
                cout << "Oops, you are allowed to take only adult Fiction books" << endl;
            }
        }

};

int main()
{
    KidUsers kid1;
    kid1.age = 10;
    kid1.bookType = "Kids";
    kid1.registerAccount();
    kid1.requestBook();

    KidUsers kid2;
    kid2.age = 18;
    kid2.bookType = "Fiction";
    kid2.registerAccount();
    kid2.requestBook();

    AdultUsers adult1;
    adult1.age = 5;
    adult1.bookType = "Kids";
    adult1.registerAccount();
    adult1.requestBook();

    AdultUsers adult2;
    adult1.age = 23;
    adult1.bookType = "Fiction";
    adult1.registerAccount();
    adult1.requestBook();

    return 0;
}

File: Week_1/OOPS/library.java
abstract class LibraryUser {
    abstract void registerAccount();
    abstract void requestBook();
}

class KidUsers extends LibraryUser {
    private int age;
    private String bookType;

    public KidUsers(int age, String bookType) {
        this.age = age;
        this.bookType = bookType;
    }

    @Override
    public void registerAccount() {
        if (age <= 12) {
            System.out.println("You have successfully registered under a Kids Account");
        } 
        else {
            System.out.println("Sorry, Age must be less than 12 to register as a kid");
        }
    }

    @Override
    public void requestBook() {
        if (bookType.equals("Kids")) {
            System.out.println("Book Issued successfully, please return the book within 10 days");
        } 
        else {
            System.out.println("Oops, you are allowed to take only kids books");
        }
    }
}

class AdultUsers extends LibraryUser {
    private int age;
    private String bookType;

    public AdultUsers(int age, String bookType) {
        this.age = age;
        this.bookType = bookType;
    }

    @Override
    public void registerAccount() {
        if (age > 12) {
            System.out.println("You have successfully registered under an Adult Account");
        } 
        else {
            System.out.println("Sorry, Age must be greater than 12 to register as an adult");
        }
    }

    @Override
    public void requestBook() {
        if (bookType.equals("Fiction")) {
            System.out.println("Book issued successfully, please return the book within 7 days");
        } 
        else {
            System.out.println("Oops, you are allowed to take only adult Fiction books");
        }
    }
}

public class library {
    public static void main(String[] args) {

        LibraryUser kid1 = new KidUsers(10, "Kids");
        kid1.registerAccount();
        kid1.requestBook();

        LibraryUser kid2 = new KidUsers(18, "Fiction");
        kid2.registerAccount();
        kid2.requestBook();

        LibraryUser adult1 = new AdultUsers(5, "Kids");
        adult1.registerAccount();
        adult1.requestBook();

        LibraryUser adult2 = new AdultUsers(23, "Fiction");
        adult2.registerAccount();
        adult2.requestBook();
    }
}